### PR TITLE
Compatibility with external deep proxies

### LINKED ISSUE #35: JSONPatcherProxy is incombatible with external deep proxies
### Steps to reproduce

1. Have an object that has a deep structure
2. Replace a subbranch in that object with a simple proxied version of that object

### Expected result

JSONPatcherProxy should observe changes to that object as usual, i.e. with the same behavior as in the initial object.

### Actual result

Some traps are set twice, also the replacement of the proxified object is detected as a replace operation.

I have an idea how to solve it using `Symbol` instead of `Map`. I am preparing a fix for that.

### PR DESCRIPTION
**Pull request by warpech**
Fixes #35 through refactoring of the metadata access layer and adding a new check to the trap of "set".

### Before this PR

```julia
jsonpatcherproxy generate operation
 Ops/sec: 95630 ±4.20% Ran 7111 times in 0.074 seconds.
jsonpatch generate operation
 Ops/sec: 158564 ±4.14% Ran 9171 times in 0.058 seconds.
jsonpatcherproxy mutation - huge object
 Ops/sec: 839 ±4.87% Ran 56 times in 0.067 seconds.
jsonpatch mutation - huge object
 Ops/sec: 989 ±2.64% Ran 55 times in 0.056 seconds.
jsonpatcherproxy generate operation - huge object
 Ops/sec: 869 ±4.87% Ran 56 times in 0.064 seconds.
jsonpatch generate operation - huge object
 Ops/sec: 980 ±2.79% Ran 55 times in 0.056 seconds.
PROXIFY big object
 Ops/sec: 1671 ±4.91% Ran 109 times in 0.065 seconds.
DIRTY-OBSERVE big object
 Ops/sec: 1945 ±2.70% Ran 109 times in 0.056 seconds.
```

### After this PR

```julia
jsonpatcherproxy generate operation
 Ops/sec: 89850 ±4.54% Ran 6729 times in 0.075 seconds.
jsonpatch generate operation
 Ops/sec: 155485 ±3.85% Ran 9117 times in 0.059 seconds.
jsonpatcherproxy mutation - huge object
 Ops/sec: 832 ±5.00% Ran 57 times in 0.069 seconds.
jsonpatch mutation - huge object
 Ops/sec: 988 ±2.60% Ran 55 times in 0.056 seconds.
jsonpatcherproxy generate operation - huge object
 Ops/sec: 807 ±5.01% Ran 53 times in 0.066 seconds.
jsonpatch generate operation - huge object
 Ops/sec: 979 ±2.74% Ran 55 times in 0.056 seconds.
PROXIFY big object
 Ops/sec: 1666 ±5.03% Ran 111 times in 0.067 seconds.
DIRTY-OBSERVE big object
 Ops/sec: 1932 ±3.28% Ran 109 times in 0.056 seconds.
```

### CODE REVIEW

File: src/jsonpatcherproxy.js [hunk 1]
```diff
@@ -276,13 +297,12 @@ const JSONPatcherProxy = (function() {
       const message =
         "You're accessing an object that is detached from the observedObject tree, see https://github.com/Palindrom/JSONPatcherProxy#detached-objects";
 
-      treeMetadata.handler.set = (
+      treeMetadata.handler.get = (
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: Why is this change needed?

File: src/jsonpatcherproxy.js [hunk 2]
```diff
@@ -193,10 +211,13 @@ const JSONPatcherProxy = (function() {
     * @constructor
     */
   function JSONPatcherProxy(root, showDetachedWarning) {
+    /**
+     * Use tree[this._metadataSymbol] to access the tree metadata when all you have is a reference to a proxified version of the tree.
+     */
+    this._metadataSymbol = Symbol("Symbol for getting the tree metadata from external access.");
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: Why it's a new symbol per instance? Can'w we create one Symbol for all instances?
Is it to be able to use JSONProxyPatcher on same object twice, and get notified twice?

File: src/jsonpatcherproxy.js [hunk 3]
```diff
@@ -34,73 +39,89 @@ const JSONPatcherProxy = (function() {
...
-    but in fact is is a proxified object moved around the tree
-    */
-    if (subtreeMetadata && !instance._isProxifyingTreeNow) {
-      subtreeMetadata.inherited = true;
+    
+    if (isObject(newValue)) {
+      const subtreeMetadata = newValue[instance._metadataSymbol];
+      if (subtreeMetadata) {
+        if(subtreeMetadata.parent === tree && subtreeMetadata.keyInParent === key) {
+          //this is the same object that we already proxified, proxified now by someone else. In this case, remain silent
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: "we" means this instance, not JSONPatcherProxy in general?

Is it necessary "proxified now by someone else" Isnt; it run also for `tree.key = tree.key`?

File: src/jsonpatcherproxy.js [hunk 4]
```diff
@@ -256,7 +277,7 @@ const JSONPatcherProxy = (function() {
     initial process;
     */
     this.pause();
-    this._isProxifyingTreeNow = true;
+    this._isProxifyingTreeNow = true; //TODO probably not needed, when I comment this out, all tests pass
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: I think https://github.com/Palindrom/JSONPatcherProxy/blob/c461b492a4fa2e2869ba99cbe0329db3846e93cd/src/jsonpatcherproxy.js#L95 exaplins it, but we just lack tests for removing moved obj.

File: src/jsonpatcherproxy.js [hunk 5]
```diff
@@ -34,73 +39,89 @@ const JSONPatcherProxy = (function() {
...
   }
+  /**
+   * A callback to be used as the proxy get trap callback.
+   * It allows to check get the tree metadata from within the tree. This allows to check if the object is proxified by this
+   * instance of JSONPatcherProxy, even when all you have is a reference that is someone else's proxy
+   * @param {JSONPatcherProxy} instance JSONPatcherProxy instance
+   * @param {Object} tree the affected object
+   * @param {String} key the effect property's name
+   */
+  function trapForGet(instance, tree, key) {
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: It would be nice to add a test to benchmarks to check how much we slowdown regular work with proxified object. As I suspect adding a trap for getting any property in entire tree could affect performance.

**Comment by tomalec** [comment]
Could you add a test for creating external deep proxy?
I still don't see a clear reason why previous implementation with a map is not capable of doing so. What's the difference between of having instance specific map than instance specific symbol?

File: src/jsonpatcherproxy.js (hunk 1 above)
Comment (warpech) [reply src/jsonpatcherproxy.js]: If you look at the `master` branch, you will see that by mistake the `set` handler was repeated twice in this file: https://github.com/Palindrom/JSONPatcherProxy/blob/a585de9f58322b1f7a770cea61f50878f26ef728/src/jsonpatcherproxy.js#L279-L294

I overlooked it in my refactor PR #36. If I saw it later, I would remove one of the `set` getters.

So the real change here was to add a new `get` handler, but I combined this with the removal of the redundant `set` handler.

File: src/jsonpatcherproxy.js (hunk 2 above)
Comment (warpech) [reply src/jsonpatcherproxy.js]: It is to be able to use JSONPatcherProxy twice on the same object, and be sure that each instance can keep its own metadata for its traps.

However, there is no test for such usage.

File: src/jsonpatcherproxy.js (hunk 3 above)
Comment (warpech) [reply src/jsonpatcherproxy.js]: >"we" means this instance, not JSONPatcherProxy in general?

Yes, "we" means "this instance of `JSONPatcherProxy`", because tree metadata is private per instance of `JSONPatcherProxy`. 

> Is it necessary "proxified now by someone else" Isnt; it run also for tree.key = tree.key?

You are right, it will also go here if it is simply reassigned at the same place.

I will clarify in the comment.

File: src/jsonpatcherproxy.js (hunk 4 above)
Comment (warpech) [reply src/jsonpatcherproxy.js]: I have linked your comment in https://github.com/Palindrom/JSONPatcherProxy/issues/40, to remember this.

File: src/jsonpatcherproxy.js (hunk 1 above)
Comment (tomalec) [reply src/jsonpatcherproxy.js]: Then do we need a message for get? `get` would not trigger any patch anyway, even if it was attached to observed object.

File: src/jsonpatcherproxy.js (hunk 2 above)
Comment (tomalec) [reply src/jsonpatcherproxy.js]: I think we need such test, to prove this change is needed and make sure the functionality will survive ;)

**Comment by warpech** [comment]
> Could you add a test for creating external deep proxy?

There is a test in the commit c461b492a4fa2e2869ba99cbe0329db3846e93cd

> I still don't see a clear reason why previous implementation with a map is not capable of doing so. What's the difference between of having instance specific map than instance specific symbol?

Consider this. There are two ways of accessing an object:
- by reference. In JSONPatcherProxy, such way of accessing is used whenever you see `tree`. This way gives you original unproxified instance of the object, because that's how the `set` trap works.
- by parent and key. In JSONPatcherProxy, such way of accessing is used whenever you see `tree[key]`. This way gives you whatever the external world sees as the value, including the proxy made by the current instance of JSONPatcherProxy and any other proxy.

All places in JSONPatcherProxy where there was a need to use `tree[key]`, used to use `tree[key]` as the key to a `Map` object. However, this only worked as long as the value at `tree[key]` was exactly the proxy object created by the current instance of JSONPatcherProxy. It the value was additionaly proxified by external code, it was detected by JSONPatcherProxy as a totally new object.

This is because in the ES Proxy world, the proxy instances have different identity than the original object:

```
var obj = {};
var p1 = new Proxy(obj, {});
var p2 = new Proxy(p1, {});
obj === p1; //false
p1 === p2; //false
```

The method used in this PR works by recognizing the object already proxified by an instance of JSONPatcherProxy not by keeping a reference to it, but by marking it from the inside:

```
var obj = {};
var meta = {};
var symb = Symbol("Symbol for getting the tree metadata from external access.");
obj[symb] = meta;

var p1 = new Proxy(obj, {});
var p2 = new Proxy(p1, {});
obj[symb] === p1[symb]; //true
p1[symb] === p2[symb]; //true
```

File: src/jsonpatcherproxy.js (hunk 5 above)
Comment (warpech) [reply src/jsonpatcherproxy.js]: Good idea, I will add such test.

**Comment by tomalec** [comment]
> There is a test in the commit c461b49

Sorry my bad

----

Thanks, now I get why having Map of object references will not work.
But then doesn't keeping `parent` object reference shares the same problem? If it doens't maybe we could keep a map of `parent, key` tuples.

I'm a little bit picky, as I remember we used to have implementation that was using `get` trap to identify proxified object, and we discarded it for some reason (I suspect performance)

**Comment by warpech** [comment]
You are right, this PR introduces about 25% performance degradation in reads. We should care about this, because performance in reads is already terrible compared to lack of proxy (see full results below).

Anyway, I think that a fix for the problem #35 is absolutely necessary. Do you have any other suggestions?

#### Before this PR

```julia
Observe and generate, small object (JSONPatcherProxy)
 Ops/sec: 94422 ±2.81% Ran 5965 times in 0.063 seconds.
Observe and generate (JSONPatcherProxy)
 Ops/sec: 1924 ±2.29% Ran 190 times in 0.099 seconds.
Primitive mutation (JSONPatcherProxy)
 Ops/sec: 480043 ±2.65% Ran 27032 times in 0.056 seconds.
Complex mutation (JSONPatcherProxy)
 Ops/sec: 6996 ±5.30% Ran 421 times in 0.060 seconds.
Serialization (JSONPatcherProxy)
 Ops/sec: 1252 ±0.77% Ran 66 times in 0.053 seconds.
```

#### After this PR

```julia
Observe and generate, small object (JSONPatcherProxy)
 Ops/sec: 94633 ±2.61% Ran 5925 times in 0.063 seconds.
Observe and generate (JSONPatcherProxy)
 Ops/sec: 2010 ±2.06% Ran 194 times in 0.097 seconds.
Primitive mutation (JSONPatcherProxy)
 Ops/sec: 414442 ±1.92% Ran 22812 times in 0.055 seconds.
Complex mutation (JSONPatcherProxy)
 Ops/sec: 6106 ±4.07% Ran 350 times in 0.057 seconds.
Serialization (JSONPatcherProxy)
 Ops/sec: 935 ±0.82% Ran 49 times in 0.052 seconds.
```

#### Full results after this PR

```julia
Observe and generate, small object (noop)
 Ops/sec: 12577915 ±0.20% Ran 641041 times in 0.051 seconds.
Observe and generate, small object (JSONPatcherProxy)
 Ops/sec: 92129 ±2.62% Ran 5823 times in 0.063 seconds.
Observe and generate, small object (fast-json-patch)
 Ops/sec: 69657 ±1.80% Ran 3888 times in 0.056 seconds.
Observe and generate (noop)
 Ops/sec: 160466 ±0.26% Ran 8364 times in 0.052 seconds.
Observe and generate (JSONPatcherProxy)
 Ops/sec: 1958 ±1.39% Ran 236 times in 0.121 seconds.
Observe and generate (fast-json-patch)
 Ops/sec: 1698 ±1.10% Ran 93 times in 0.055 seconds.
Primitive mutation (noop)
 Ops/sec: 2549711 ±0.60% Ran 130739 times in 0.051 seconds.
Primitive mutation (JSONPatcherProxy)
 Ops/sec: 403460 ±3.16% Ran 22659 times in 0.056 seconds.
Primitive mutation (fast-json-patch)
 Ops/sec: 7835 ±0.20% Ran 401 times in 0.051 seconds.
Complex mutation (noop)
 Ops/sec: 6551384 ±0.20% Ran 331525 times in 0.051 seconds.
Complex mutation (JSONPatcherProxy)
 Ops/sec: 6035 ±6.15% Ran 354 times in 0.059 seconds.
Complex mutation (fast-json-patch)
 Ops/sec: 7707 ±0.62% Ran 406 times in 0.053 seconds.
Serialization (noop)
 Ops/sec: 7206 ±0.41% Ran 375 times in 0.052 seconds.
Serialization (JSONPatcherProxy)
 Ops/sec: 948 ±0.35% Ran 50 times in 0.053 seconds.
Serialization (fast-json-patch)
 Ops/sec: 7101 ±0.40% Ran 373 times in 0.053 seconds.
```

**Comment by tomalec** [comment]
Have you tried 
> But then doesn't keeping parent object reference shares the same problem? If it doesn't maybe we could keep a map of parent, key tuples.

**Comment by tomalec** [comment]
I agree it's necessary to fix that problem over the performance, but we should do our best not to degrade it too much. Therefore if the idea of keeping a map of parent+key does not work, let's merge it.

File: src/jsonpatcherproxy.js (hunk 1 above)
Comment (tomalec) [reply src/jsonpatcherproxy.js]: @warpech ping

File: test/spec/proxyBenchmark.js [hunk 6]
```diff
@@ -58,69 +43,153 @@ function generateBigObjectFixture(carsSize) {
...
+}
+
+/* ============================= */
+
+{
+  const suiteName = 'Observe and generate, small object';
+  
+  {
+    suite.add(`${suiteName} (noop)`, function() {
+      const obj = generateBigObjectFixture(1);
```
Comment (tomalec) [inline test/spec/proxyBenchmark.js]: Shouldn't generation of obj be moved outside of suite? (here and bellow?)

Comment (tomalec) [reply test/spec/proxyBenchmark.js]: Ah, it shouldn't as we want to proxify new object every time..

File: test/spec/proxyBenchmark.js [hunk 7]
```diff
@@ -58,69 +43,153 @@ function generateBigObjectFixture(carsSize) {
...
+      modifyObj(obj);
+    });
+  }
+
+  {
+    suite.add(`${suiteName} (JSONPatcherProxy)`, function() {
+      const obj = generateBigObjectFixture(100);
+      const jsonPatcherProxy = new JSONPatcherProxy(obj);
+      const observedObj = jsonPatcherProxy.observe(true);
+      const patches = jsonPatcherProxy.generate();
```
Comment (tomalec) [inline test/spec/proxyBenchmark.js]: I think you should generate patches after you make changes,

File: test/spec/proxyBenchmark.js [hunk 8]
```diff
@@ -58,69 +43,153 @@ function generateBigObjectFixture(carsSize) {
...
+    });
+  }
+
+  {
+    const obj = generateBigObjectFixture(100);
+    const jsonPatcherProxy = new JSONPatcherProxy(obj);
+    const observedObj = jsonPatcherProxy.observe(true);
+    
+    suite.add(`${suiteName} (JSONPatcherProxy)`, function() {
+      observedObj.cars[50].name = reverseString(observedObj.cars[50].name);
```
Comment (tomalec) [inline test/spec/proxyBenchmark.js]: There is no `generate` call as in `fast-json-patch`

File: test/spec/proxyBenchmark.js [hunk 9]
```diff
@@ -58,69 +43,153 @@ function generateBigObjectFixture(carsSize) {
...
+  }
+
+  {
+    const obj = generateBigObjectFixture(100);
+    const jsonPatcherProxy = new JSONPatcherProxy(obj);
+    const observedObj = jsonPatcherProxy.observe(true);
+  
+    suite.add(`${suiteName} (JSONPatcherProxy)`, function() {
+      const item = observedObj.cars.shift();
+      observedObj.cars.push(item);
```
Comment (tomalec) [inline test/spec/proxyBenchmark.js]: There is no `generate` call

File: test/spec/proxyBenchmark.js (hunk 7 above)
Comment (warpech) [reply test/spec/proxyBenchmark.js]: I didn't change this. It is how it works on `master`. But I think you are right, I will change it. 

@alshakero do you remember why did you chose to call `generate` here not later?

File: test/spec/proxyBenchmark.js (hunk 8 above)
Comment (warpech) [reply test/spec/proxyBenchmark.js]: I didn't change this. It is how it works on `master`. But I think you are right, I will change it. 

@alshakero do you remember why did you chose to call `generate` here not later?

File: test/spec/proxyBenchmark.js (hunk 9 above)
Comment (warpech) [reply test/spec/proxyBenchmark.js]: I didn't change this. It is how it works on `master`. But I think you are right, I will change it. 

@alshakero do you remember why did you chose to call `generate` here not later?

**Comment by warpech** [comment]
I fixed the `generate` calls. Current numbers:

#### After this PR

```julia
Observe and generate, small object (JSONPatcherProxy)
 Ops/sec: 95913 ±2.54% Ran 5936 times in 0.062 seconds.
Observe and generate (JSONPatcherProxy)
 Ops/sec: 2008 ±1.49% Ran 210 times in 0.105 seconds.
Primitive mutation (JSONPatcherProxy)
 Ops/sec: 436081 ±0.48% Ran 22251 times in 0.051 seconds.
Complex mutation (JSONPatcherProxy)
 Ops/sec: 7526 ±0.37% Ran 387 times in 0.051 seconds.
Serialization (JSONPatcherProxy)
 Ops/sec: 916 ±0.29% Ran 47 times in 0.051 seconds.
```

The numbers acutally went up, because before this commit, the queue of recorded patches was growing infinitely. Now, the queue is flushed after each repetition.

File: test/spec/proxyBenchmark.js (hunk 6 above)
Comment (warpech) [reply test/spec/proxyBenchmark.js]: > Ah, it shouldn't as we want to proxify new object every time..

Yes

File: test/spec/proxyBenchmark.js (hunk 9 above)
Comment (alshakero) [reply test/spec/proxyBenchmark.js]: When it comes to JSONPatcherProxy, `generate` call isn't needed here, or anywhere to be honest. All the heavy-lifting occurs synchronously upon the mutation itself. I think it's safe to remove all `generate` calls.

File: test/spec/proxyBenchmark.js (hunk 8 above)
Comment (alshakero) [reply test/spec/proxyBenchmark.js]: See https://github.com/Palindrom/JSONPatcherProxy/pull/39/files/bcde61c41f9ebe5487fe4a5c712bb2d911a064e9..67cc6f570e275fcb4e1c9a886eb9977e3421bf83#r291562016

File: test/spec/proxyBenchmark.js (hunk 7 above)
Comment (alshakero) [reply test/spec/proxyBenchmark.js]: I think line 106 can be removed altogether. See https://github.com/Palindrom/JSONPatcherProxy/pull/39/files/bcde61c41f9ebe5487fe4a5c712bb2d911a064e9..67cc6f570e275fcb4e1c9a886eb9977e3421bf83#r291562016

Comment (alshakero) [reply test/spec/proxyBenchmark.js]: Ah that would be a leak as shown in @warpech comment https://github.com/Palindrom/JSONPatcherProxy/pull/39#issuecomment-499857818

File: test/spec/proxyBenchmark.js (hunk 9 above)
Comment (warpech) [reply test/spec/proxyBenchmark.js]: Since the heavy-lifting occurs synchrously, `generate` has low cost but makes the test symmetrical with the test done using `fast-json-patch`. Hence, I think it is good to keep `generate` in both places

File: src/jsonpatcherproxy.js (hunk 1 above)
Comment (tomalec) [reply src/jsonpatcherproxy.js]: ok, my idea was simply wrong and would just introduce more complexity to previous buggy solution.

**Comment by warpech** [comment]
I pushed an additional test to check if double proxification by `JSONPatcherProxy` works as intended. It does work.

**Comment by warpech** [comment]
I am not merging this until we find it absolutely necessary for view binding.

One thing that I find very annoyning about this change is how it affects debugging. Any attempt to read a property from the observed object causes the debugger to step into the `trapForGet`.

**Comment by warpech** [comment]
I made a new PR that contains only the test refactorings from the current PR: https://github.com/Palindrom/JSONPatcherProxy/pull/42

**Comment by warpech** [comment]
This PR was rebased on current `master` and is now available as a new PR #43
